src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java

```diff
@@ -50,347 +53,392 @@
 public class PermissionServiceBean {
 
     private static final Logger logger = Logger.getLogger(PermissionServiceBean.class.getName());
-    
+
+    /**
+     * A set of permissions that con only be granted to {@link AuthentictedUser}s.
+     */
     private static final Set<Permission> PERMISSIONS_FOR_AUTHENTICATED_USERS_ONLY = 
-           EnumSet.copyOf(Arrays.asList(Permission.values()).stream()
-                    .filter( Permission::requiresAuthenticatedUser )
-                    .collect( Collectors.toList() ));
-    
+        EnumSet.copyOf(Arrays.asList(Permission.values()).stream()
+                    .filter(Permission::requiresAuthenticatedUser)
+                    .collect(Collectors.toSet()));
+
     @EJB
     BuiltinUserServiceBean userService;
-    
+
     @EJB
     AuthenticationServiceBean authenticationService;
 
     @EJB
     DataverseRoleServiceBean roleService;
 
     @EJB
-    RoleAssigneeServiceBean roleAssigneeService;    
-    
+    RoleAssigneeServiceBean roleAssigneeService;
+
     @EJB
     DataverseServiceBean dataverseService;
+    
+    @EJB
+    DvObjectServiceBean dvObjectServiceBean;
 
     @PersistenceContext
     EntityManager em;
 
     @EJB
     GroupServiceBean groupService;
-    
+
     @Inject
     DataverseSession session;
-    
+
     @Inject
     DataverseRequestServiceBean dvRequestService;
     
+    
     /**
-     * A request-level permission query (e.g includes IP groups).
+     * This interface is used locally by {@link EjbDataverseEngine} beans to
+     * create detailed reports on which permissions are missing. Probably not
+     * useful beyond that area.
+     */
+    interface InsufficientPermissionCallback {
+        void insufficientPermissions( DvObject dvo, Set<Permission> required, Set<Permission> granted);
+    }
+    
+    /**
+     * A request-level permission query (e.g includes IP groups). These queries
+     * form a DSL for querying permissions, in a way that's hopefully more
```

Comment:
What's a DSL?

Reply:
Domain Specific Language. Updated the docs.